src/TestInternalGrains/PersistenceTestGrains.cs

```diff
@@ -500,6 +501,22 @@ public override Task OnActivateAsync()
             return base.OnActivateAsync();
         }
 
+        // When reentrant grain is doing WriteStateAsync, etag violations are posssible due to concurent writes.
+        // The solution is to serialize all writes, and make sure only a single write is outstanind at any moment in time.
+        // No deadlocks are posssible with that approach, since all WriteStateAsync go to the store, which does not issue call into grains,
+        // thus cycle of calls is not posssible.
+        // Implementaton: need to use While and not if, due to the same "early check becomes later invalid" standard problem, like in conditional variables.
+        private async Task PerformSerializedStateUpdate()
+        {
+            while (outstandingWriteStateOperation != null)
+            {
+                await outstandingWriteStateOperation;
+            }
+            outstandingWriteStateOperation = WriteStateAsync();
+            await outstandingWriteStateOperation;
+            outstandingWriteStateOperation = null;
```

Comment:
this *might* work depending on how the scheduler schedules work, but is a little bit flaky and prone to disruptions if we optimize differently to execute inline if a task is completed. The reasoning is that the request caught in the while loop might wake up before the `outstandingWriteStateOperation = null` line. If it does, then the while condition will still be true, and call `await outstandingWriteStateOperation`, but this time the Task is completed. If that executes inline, then it will be caught in a synchronous loop not allowing the other request to null out the field.
There were also several other issues in the case of exceptions being thrown from the call to base.WriteStateAsunc(), and potentially break that grain activation forever.

I proposed another solution in PR #1750, and if we like it, we could do it here too.